Support "Select * from (a)" queries (when the table name is inside Parenthesis)

eyalleshem opened the pull request:
Following the discussion in issue #154  , 
I tried to add support to parenthesis around table name, 
waiting for reviews :)

nickolay commented on src/parser.rs:
```diff
@@ -1794,47 +1829,29 @@ impl Parser {
+                                TableFactor::NestedJoin { .. } => (),
+                                _ => {
+                                    if table_and_joins.joins.is_empty() {
+                                        // The SQL spec prohibits derived tables and bare
```
I expected the change to be here, as _"bare tables .. appearing alone in parentheses"_ is exactly the case you reported as supported in Snowflake SQL. Not sure about the "derived tables .. alone in parentheses" case.

eyalleshem commented:
Hi , 
From (foo bar) should fail ,  i also added test for that . 

I think added the bare table before the join - as I don't want the object to be "table with joins" but without join's   - i think it's will be more correct that bare table  will be just a table_factor (I also updated the comment's in that section) .

eyalleshem commented:
![image](https://user-images.githubusercontent.com/38852709/78252112-0227d680-74fb-11ea-9827-96559b51f55a.png)
It's look like the CI don't failed because code issue - but because some infrastructure issue..
Any option to re-run it ?

nickolay commented:
> It's look like the CI don't failed because code issue - but because some infrastructure issue..
> Any option to re-run it ?

I don't know how travis permissions work, but it allows me to restart the build once I log in with github. Judging from https://rust-lang-nursery.github.io/rust-toolstate/ the nightly rustfmt we use is currently unavailable, so restarting the build until that's fixed will not help.

nickolay commented:
> I think added the bare table before the join - as I don't want the object to be "table with joins" but without join's - i think it's will be more correct that bare table will be just a table_factor 

The joins in `TableWithJoins` are optional by design. A more appropriate name could be `TableFactorOptionallyFollowedByJoins`, but that's too long.

The code you have now strips the extraneous parens, while it would arguably be more correct to keep them in the AST via `TableFactor::NestedJoin` -- probably renamed it to `Nested` and its comments updated https://github.com/andygrove/sqlparser-rs/blob/172ba42001003bce79708141923ba6100cc7d4f2/src/ast/query.rs#L228-L232

I believe this would be less code, less nesting, and less intrusive patch overall.

eyalleshem commented:
But - if i understand right you suggestion : 

"From a" - will be parsed as  : 
```
TableWithJoins  {
    relation :  TableFactor::Table {
        name :a 
      [...] 
   }
}
```

"From (a)" - will be parsed as  : 
```
TableWithJoins  {
    relation :  TableFactor::NestedJoin (
        TableWithJoins  {
              [Table factor for a]
         }
   ) 
}
```

but we don't have any nested join in the query...

nickolay commented:
Right. `TableFactor::NestedJoin` was called that because standard SQL only allows joins to be nested. If we want other things to be nested, it should _"probably renamed it to Nested and its comments updated," as I noted._

You may argue that an object representing the extra parentheses does NOT belong to an _abstract_ syntax tree, but in other cases we prefer preserving detail in the AST, and in this particular case IMO this should make the particularly complex piece of the parser easier to manage.

eyalleshem commented:
Changed according to your suggestion. 
It's also will be better - as it's will support multi level of parenthesis. 
(select *from ((a)))

nickolay commented on src/parser.rs:
```diff
@@ -1793,18 +1795,12 @@ impl Parser {
                     // Ignore the error and back up to where we were before.
                     // Either we'll be able to parse a valid nested join, or
```
This whole comment should be reworded (as well as the bit in ast.rs I linked to earlier), as otherwise it becomes self-contradictory. I guess I'll do it myself later.

eyalleshem replied:
Ok , i didn't touched  this comment - and according to my english skills i don't sure that I could improve it :) .

nickolay commented on tests/sqlparser_common.rs:
```diff
@@ -1806,11 +1806,34 @@ fn parse_join_nesting() {
 
-    let res = parse_sql_statements("SELECT * FROM (a NATURAL JOIN (b))");
+
+    let sql = "SELECT * FROM a NATURAL JOIN (((b NATURAL JOIN c)))";
```
This appears to duplicate the test above, no?

eyalleshem replied:
yep , my mistake - removed

nickolay commented on tests/sqlparser_common.rs:
```diff
@@ -1806,11 +1806,34 @@ fn parse_join_nesting() {
+        vec![join(nest!(nest!(nest!(table("b"), table("c")))))]
+    );
+
+    let sql = "SELECT * FROM (a NATURAL JOIN (b))";
```
It would be cleaner to have only the generic dialect accept this, and have a TODO test that makes the ANSI dialect fail it. But I guess that introducing the necessary infrastructure is too much for this PR.

So at least we must note that this test is supposed to succeed with the Snowflake dialect only:
```
// Parenthesized table names are non-standard, but supported in Snowflake SQL
```

eyalleshem replied:
Agree that dialect  will be much cleaner , Sound like small project  :) . 
Added the comment.

nickolay commented on tests/sqlparser_common.rs:
```diff
@@ -1806,11 +1806,34 @@ fn parse_join_nesting() {
+        nest!(table("a"), nest!(table("b")))
+    );
+
+    let sql = "SELECT * FROM (a NATURAL JOIN ((b)))";
```
```
// Double parentheses around table names are non-standard, but supported in Snowflake SQL
```

Please check that it is.

eyalleshem replied:
Added.

eyalleshem replied:
and checked :)

nickolay commented on tests/sqlparser_common.rs:
```diff
@@ -1953,11 +1976,26 @@ fn parse_derived_tables() {
         }))
     );
 
-    let res = parse_sql_statements("SELECT * FROM ((SELECT 1) AS t)");
```
```
// Nesting a subquery in parentheses is non-standard, but supported in Snowflake SQL
```

Please check that it is.

eyalleshem replied:
Added , thank for the help with the comment wording. 
(checked in snowflake..)

eyalleshem commented on src/parser.rs:
```diff
@@ -1767,10 +1767,12 @@ impl Parser {
+            //                   | | | |                                                |
+            //                   | | | | /----------------------------------------------|
+            //                   | | | | |  
+            //                   | | | | (5) some db's (snowflake) allow bare table inside parenthesis                                                  
```
removed

eyalleshem commented:
Hi , 
The Ci failed because some clippy warning in era that i didn't touched (examples/cli.rs:43:26), 
and i resolved all the comments  - but for some reason the PR still in state of "Change requested" - don't sure how i need to solve that..

nickolay commented:
>  The Ci failed because some clippy warning in era that i didn't touched (examples/cli.rs:43:26),

Thanks, I opened #159 to fix that.

> some reason the PR still in state of "Change requested" - don't sure how i need to solve that..

I think you should be able to expand that section and "Re-request review" to change its state.

That's not the reason I didn't get to it, though, sorry!
